### DISCUSSION TITLE
Extract potential facts from text?

### DISCUSSION THREAD

**Discussion by Chaddeus** [winkjs/wink-nlp#104] [Help on coding] [answered]
What might be a good way to analyze a passage of text to determine if there might be things an author should double check before publishing their work?

Entities looks promising, but it's returning almost too much. Some parts are perfect, but some are things like "one" -- which *could* be important to check, but is likely a false-positive.

I wonder if filters might help, but unsure how to implement it in this fashion.

Thank you.

**Comment by sanjayaksaxena** [comment, marked as answer]
Hello @Chaddeus can you share one/two representative examples to help us understand the question better... Sanjaya

**Reply by Chaddeus** [reply]
Yes, here's the entity list first:
```
[
    "one",
    "around 200",
    "500 yen",
    "nearly 8,000",
    "one"
]
```
I can easily remove duplicates. But the "one" in there is hard to show to my users as "a fact to double check" I think.

Here's most of the passage that was analyzed:

> Are you a hungry tourist in Tokyo, trying to save money while dining? Look no further than the delicious ramen options available! Before you try ramen in Tokyo, you may have been overwhelmed by the prices of restaurants and limited by your budget. But with so many types of ramen to choose from and prices ranging from as low as $3-4 per bowl, you can easily find something within your budget that tastes amazing! 
> 
> The variety of ramen available means there's something for everyone. From classic miso or shoyu broth to more unique flavors like paitan or tsukemen, there's something for anyone who loves noodles. And if you're looking to save even more money, there are plenty of budget shops that offer delicious ramen at a fraction of the price. 
> 
> Ramen in Tokyo is the perfect way to enjoy a tasty meal on a budget. With mouth-watering flavors, quality ingredients, and wallet-friendly prices, it's a great option for travelers looking to make the most out of their dining experience in Tokyo. So come and try `one` of Tokyo's best dishes - you won't regret it!
> 
> Are you a hungry tourist on a budget looking to dine in Tokyo?
> 
> Look no further than the magical world of ramen! Ramen is a beloved part of Japanese food culture, and Tokyo has some of the best ramen shops in the world.
> 
> Eating ramen in Tokyo is an experience not to be missed, and there are plenty of ways to satisfy your cravings without breaking the bank. In this paper, we'll be discussing the different ramen varieties available in Tokyo, tips for finding quality ramen on the cheap, and a few examples of the best budget-friendly ramen shops in the city.
> 
> So, if you're looking for a delicious and affordable meal in Tokyo, look no further than the delicious world of ramen!
> 
> ### Definition of Ramen
> Ramen is a delicious and budget-friendly way to enjoy a meal while touring Tokyo. For hungry tourists looking for a cheap yet satisfying option, Ramen is a great choice. Not only is it conveniently sold in convenience stores, supermarkets, and restaurants all around the city, but it also comes in a variety of flavors. From the classic miso ramen to the spicy tonkotsu ramen, there is something for everyone. Plus, the price of a single bowl of ramen ranges from `around 200` to `500 yen`, making it a great economical choice for budget-conscious diners.
> 
> And when you think about it, ramen is actually the perfect meal for traveling.
> 
> ### Reasons to Eat Ramen in Tokyo
> Eating ramen in Tokyo is a great way to save money while enjoying the renowned excellence of the city's cuisine. With `nearly 8,000` ramen restaurants in Tokyo, it's tough to make a wrong decision when it comes to finding a satisfying meal. From traditional establishments to hip, modern joints, these eateries offer a variety of delicious flavors and styles of ramen, ranging from shio (salt-based) to miso (soybean-based) to tonkotsu (pork bone-based), all at affordable prices. Plus, the fast and friendly service makes it the perfect meal for an adventurous traveler looking to explore the city's many sights and sounds.

Thank you for the quick reply. 🙏

**Reply by sanjayaksaxena** [reply]
If we run nlp pipeline on the above text
```javascript
// Assuming init code is in place & text has the above example text!
const doc = nlp.readDoc( text );
console.log( doc.entities().out( its.detail ) );
```

It produces the following output:
```
Array (4 items):
0: Object {value: "one", type: "CARDINAL"}
1: Object {value: "around 200", type: "CARDINAL"}
2: Object {value: "500 yen", type: "MONEY"}
3: Object {value: "nearly 8,000", type: "CARDINAL"}
```

What exactly you would like to achieve on this array?

**Reply by Chaddeus** [reply]
Ah... I wish maybe "one" was somehow differentiated from "around 200" -- as "around 200" is something from the text my users probably should review before publishing their content.

I suppose the only thing about "one" being different from the other CARDINAL types is that it's a single word. Would it be wrong to assume that for CARDINAL types, a single word entity is less important?

**Reply by sanjayaksaxena** [reply]
Hmm... if "one" is a specific case, then it can be easily filtered... but if you have more general need then we need to define the "need" and accordingly figure out the solution.

**Reply by Chaddeus** [reply]
Agreed.

Essentially, I want to show my users (who are writers) the potential facts in their writing. Just a way to show them and say "Hey, these are things you will want to double-check before you publish).

Not all writers are in the same niche/topic. Some might be bloggers, some might be journalists, others might be fiction authors.

For my fiction authors it's not as important, but journalists should definitely fact check before publishing their work.

I suppose I could categorize the entities into an accordion and let the writers decide if something is important to review or not. But I think if there were a lot of entities like "one" then it might become an unappealing feature of the app. 😇

**Reply by sanjayaksaxena** [reply]
One possible way to tackle is to count (i.e. frequency) such instances and if they are above a threshold then either ignore or give a special treatment; also the count can be decided on the basis of the length of the text!

Since you are helping your users to be factual, consider detecting unsupported attribution phrases such as — "often reported", "widely thought", "experts/scientists claim", "it is considered/believed", etc. using [custom entities](https://winkjs.org/wink-nlp/custom-entities.html) and provide feedback!

**Reply by Chaddeus** [reply]
Ooh, that's interesting! Thank you for the custom entities tip. That's smart.

Any other interesting ways you've seen custom entities used?

**Reply by sanjayaksaxena** [reply]
It all depends on the need & context!

**Reply by Chaddeus** [reply]
I wonder if there are some interesting way this could help with generative AI? Perhaps around logit bias, which requires delivering tokens rather than keywords. Perhaps for a different discussion. :)

**Reply by sanjayaksaxena** [reply]
Sure, we can discuss separately... Consider contributing...

For extractive summary you can look up: https://observablehq.com/@winkjs/how-to-visualize-key-sentences-in-a-document

**Reply by sanjayaksaxena** [reply]
Also [`its.readabilityStats`](https://winkjs.org/wink-nlp/its-as-helper.html) helper can be useful in your context.

**Reply by Chaddeus** [reply]
Yes, already testing readabilityStats -- super cool, nice work with that one.

And you read my mind, I was already reading that how-to visualize key sentences article. Curious what the percentage slider means... and how "importance" is considered. Relative length?

**Reply by sanjayaksaxena** [reply]
Slider controls the %age of sentences to be highlighted & color intensity is relative importance.

**Reply by Chaddeus** [reply]
Interesting. What magic is going on to determine one sentence has importance over others in the passage?

**Reply by sanjayaksaxena** [reply]
See `// Compute relative importance of each sentence as a numerical value between 0 and 1.` section there...

**Reply by Chaddeus** [reply]
Nice. Is `sentenceWiseNormalizedWeights` based on number of words in the sentence? Is there any sort of analysis of content/context happening to determine "importance?"

**Reply by sanjayaksaxena** [reply]
It is not on the number of words in the sentence... please go thru the logic.

**Reply by Chaddeus** [reply]
Ah, I see it now. Interesting. Thank you.

**Comment by rachnachakraborty** [comment]
Hi @Chaddeus,

Would be wonderful to mark this thread as answered if discussion has been helpful to you.

Your focus on NLP based features makes us curious about your application, would be happy to learn more about it.

Looking forward to hearing from you.

Best,
Rachna

**Reply by Chaddeus** [reply]
Oh, oops. I thought this was a "discussion" - didn't realize there was a "mark as answer" toggle.
